src/ByteCountingStream.php

```diff
@@ -0,0 +1,63 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ *  Stream decorator that ensures an expected number of bytes can be read
+ *  from an underlying read stream. \RuntimeException is thrown when
+ *  the underlying stream fails to provide the expected number of bytes.
+ *  Excess bytes will be ignored.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes remains to be read */
+    private $remaining;
+
+    /**
+     * @param StreamInterface $stream       Stream to wrap
+     * @param int             $bytesToRead  Number of bytes to read
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(StreamInterface $stream, $bytesToRead)
+    {
+        $this->stream = $stream;
+
+        if (!is_int($bytesToRead) || $bytesToRead < 0) {
+            $msg = "Bytes to read should be non-negative integer, got {$bytesToRead}.";
+            throw new \InvalidArgumentException($msg);
+        }
+
+        if (
+            $this->stream->getSize() !== null &&
+            $bytesToRead > $this->stream->getSize()
+        ) {
+            $msg = "The ByteCountingStream decorator expects to be able to "
+                . "read {$bytesToRead} from a stream, but the stream being decorated "
+                . "only contains {$this->stream->getSize()} bytes.";
+            throw new \InvalidArgumentException($msg);
```

Comment:
I feel like this should be a RuntimeException because it's essentially the same failure mode as reach EOF before reading the target number of bytes.

I'm also thinking that perhaps this should have its own Exception class. This is going to be an exception that developers may way to actually handle explicitly, and just throwing RuntimeException might not be good enough in this particular case. I know we use RuntimeException in a lot of places in this project, but this error case seems like something you'd definitely be able to anticipate and recover from, thus I see more of a need for an explicit exception.... What do you think?

@Tobion What do you think about the PR and whether or not here should be an explicit exception for this failure case and the failure case encountered in the decorated read() method?